lib/Target/LLVM/LLVMToDynamicLib.cpp

```diff
@@ -0,0 +1,313 @@
+  std::string errorMessage;
+  auto llvmTarget =
+      llvm::TargetRegistry::lookupTarget(targetTriple, errorMessage);
+  if (!llvmTarget) {
+    llvm::errs() << "target lookup failed for " << targetTriple
+                 << " w msg: " << errorMessage << "\n";
+    return nullptr;
+  }
+
+  llvm::TargetOptions options;
+
+  std::unique_ptr<llvm::TargetMachine> machine(llvmTarget->createTargetMachine(
+      targetTriple, "generic" /* cpu e.g k8 */,
+      "" /* cpu features e.g avx512f */, options, llvm::Reloc::Model::PIC_));
+  return machine;
+}
+
+llvm::LogicalResult compileToObject(llvm::Module &module,
+                                    llvm::LLVMContext &context,
+                                    llvm::StringRef outputFilename) {
+
+  //  Initialize LLVM targets
+  // TODO (vwells): eventually, we should get this working on other archs, but
+  // adding new archs requires corresponding cmake changes
+  LLVMInitializeX86Target();
+  LLVMInitializeX86TargetMC();
+  LLVMInitializeX86TargetInfo();
+  LLVMInitializeX86AsmPrinter();
+  LLVMInitializeX86AsmParser();
+
+  // Set target triple if not already set
+  if (module.getTargetTriple().empty()) {
+    auto defaultTriple = llvm::sys::getDefaultTargetTriple();
+    module.setTargetTriple(defaultTriple);
+  }
+
+  auto targetMachine = createTargetMachine(module.getTargetTriple());
+  if (!targetMachine) {
+    llvm::errs() << "Failed to create TargetMachine for triple: "
+                 << module.getTargetTriple() << "\n";
+    return llvm::failure();
+  }
+
+  // Set data layout
+  module.setDataLayout(targetMachine->createDataLayout());
+
+  // Create an output file stream to write the object file
+  std::error_code EC;
+  llvm::raw_fd_ostream out(outputFilename, EC, llvm::sys::fs::OF_None);
+  if (EC) {
+    llvm::errs() << "Error opening output file: " << EC.message() << "\n";
+    return llvm::failure();
+  }
+
+  // Emit object code to the file
+  llvm::legacy::PassManager passManager;
+  passManager.add(
+      new llvm::TargetLibraryInfoWrapperPass(targetMachine->getTargetTriple()));
+  if (targetMachine->addPassesToEmitFile(passManager, out, nullptr,
+                                         llvm::CodeGenFileType::ObjectFile)) {
+    llvm::errs() << "Target machine cannot emit object file\n";
+    return llvm::failure();
+  }
+
+  passManager.run(module);
+
+  return llvm::success();
+}
+
+// simple wrapper for running link command + error handling
+llvm::LogicalResult runLinkCommand(llvm::StringRef commandLine) {
+  llvm::dbgs() << "Running linker command:\n" << commandLine << "\n";
+  const auto exitCode = system(commandLine.data());
+  if (exitCode == 0) {
+    return llvm::success();
+  }
+  llvm::errs() << "Linking failed; escaped command line returned exit code "
+               << exitCode << ":\n\n"
+               << commandLine << "\n\n";
+  return llvm::failure();
+}
+
+// TODO: decide if we have any use cases where we actually need multiple .o
```

Comment:
TODO need issue number.

Reply:
this one seems like a silly issue to me, so I'm just deleting the comment.  I think supporting multiple .o files in case we need in future is fine